File: backend/app/services/anonymous_admin.py
"""Telegram-like anonymous admin posting helpers."""

from __future__ import annotations

from typing import Optional, Tuple


def resolve_anonymous_sender_display(
    *,
    is_anonymous: bool,
    group_title: Optional[str],
    real_sender_name: Optional[str],
    viewer_is_sender: bool,
    viewer_is_admin: bool,
) -> Tuple[Optional[str], bool]:
    """
    Returns (display_name, is_anonymous_flag_for_client).

    Members see the group title. Sender/admins see "Title (Real Name)".
    """
    if not is_anonymous:
        return real_sender_name, False
    title = (group_title or "").strip() or "Группа"
    if viewer_is_sender or viewer_is_admin:
        real = (real_sender_name or "").strip() or "Админ"
        return f"{title} ({real})", True
    return title, True


def can_send_anonymously(*, is_group: bool, is_admin: bool) -> bool:
    return bool(is_group and is_admin)

File: lib/features/chat/application/anonymous_admin.dart
// Telegram-like anonymous admin display helpers.

bool canSendAnonymously({
  required bool isGroup,
  required bool amIGroupAdmin,
}) {
  return isGroup && amIGroupAdmin;
}

/// Members see [groupTitle]; sender/admins see "Title (Real Name)".
String? resolveAnonymousSenderLabel({
  required bool isAnonymous,
  required String? groupTitle,
  required String? realSenderName,
  required bool viewerIsSender,
  required bool viewerIsAdmin,
}) {
  if (!isAnonymous) return realSenderName;
  final title = (groupTitle ?? '').trim();
  final safeTitle = title.isEmpty ? 'Группа' : title;
  if (viewerIsSender || viewerIsAdmin) {
    final real = (realSenderName ?? '').trim();
    final safeReal = real.isEmpty ? 'Админ' : real;
    return '$safeTitle ($safeReal)';
  }
  return safeTitle;
}
